src/mon/PGMap.cc

```diff
@@ -3546,10 +3546,14 @@ int process_pg_map_command(
     string val;
     vector<string> dumpcontents;
     set<string> what;
+    string format;
     if (cmd_getval(cmdmap, "dumpcontents", dumpcontents)) {
       copy(dumpcontents.begin(), dumpcontents.end(),
            inserter(what, what.end()));
     }
+    cmd_getval(cmdmap, "format", format, string("plain"));
```

Comment:
instead of resetting the input parameter to the one created here. i'd suggest

1. set `f` in `DaemonServer::_handle_command()`, like
```c++
  boost::unique_ptr<Formatter> f;
  {
    std::string format;
    if (boost::algorithm::ends_with(prefix, "_json")) {
      format = "json";
    } else {
      cmd_getval(cmdctx->cmdmap, "format", format, string("plain"));
    }
    f.reset(Formatter::create(format));
  }
```

2. do not `cmd_putval(g_ceph_context, cmdmap, "format", string("json"));` in `process_pg_map_command()`.

because

1. in `DaemonServer::_handle_command()`, we have
```c++
       f->open_object_section("pg_info");
       f->dump_bool("pg_ready", pgmap_ready);
```
how can we ensure that this part is printed using the JSON format

2. we need to call `f->flush(cmdctx->odata);` to flush the text printed using `f` to `cmdctx->odata`.
how can we flush the text collected using the `f` created in `process_pg_map_command()`?

probably we should do further preprocessing in `DaemonServer::_handle_command()`. but could leave it for a follow-up change.

Reply:
reply to 2: cmd_putval(g_ceph_context, cmdmap, "format", string("json")); was already there before changes. And it doesn't seem to have any effect.

Reply:
1. DaemonServer::_handle_command only check "format" parameter in the very beginning, and `cmd_putval(g_ceph_context, cmdmap, "format", string("json"));` was already in process_pg_map_command before theses changes, and it doesn't seem to have any effect. So (maybe) use cmd_getval inside won't have bad effects (aka. make outside format become json unexpectedly).
However it's ok to remove cmd_putval if we decide to leave the format things outside in DaemonServer.
2. It looks like that cmdctx->odata was transmitted to process_pg_map_command by DaemonServer, and f->flush(*odata) is doing something like appending strings. Besides DaemonServer:921 alse defines a new Formatter and flush itself (with the same odata before).
3. Altough it's not quite elegant (i think) parsing the command line string, but since there are several places elsewere doing the same thing, ok to me, and I will change my implemention.

Reply:
> Altough it's not quite elegant (i think) parsing the command line string, but since there are several places elsewere doing the same thing, ok to me, and I will change my implemention.

agreed. please feel free to create a follow-up PR to address it.